File: sdks/python/src/memoturn/openai.py
"""Drop-in OpenAI wrapper — records chat completions and Responses API calls as
generations, including streaming (``stream=True``) calls for both endpoints."""
from __future__ import annotations

from typing import Any, Optional

from ._stream import _ABANDONED, _RecordingStream, _get
from .client import Memoturn, Trace
from .decorator import get_client


def wrap_openai(client: Any, memoturn: Optional[Memoturn] = None, *, trace: Optional[Trace] = None) -> Any:
    """Patch ``client.chat.completions.create`` and ``client.responses.create`` to trace
    calls. Returns the same client. ``responses`` is patched only if present (newer SDKs).

    Streaming calls (``stream=True``) are recorded too: the returned stream is wrapped so
    chunks are forwarded unchanged to the caller while being accumulated into the same
    output/usage shape as a non-streaming call, and the generation is closed when the
    stream is exhausted, errors, or is abandoned (closed/garbage-collected/idle-timed-out
    before completion).
    """
    mt = memoturn or get_client()
    _patch_chat(client, mt, trace)
    _patch_responses(client, mt, trace)
    return client


def _patch_chat(client: Any, mt: Memoturn, trace: Optional[Trace]) -> None:
    completions = client.chat.completions
    original = completions.create

    def create(*args: Any, **kwargs: Any) -> Any:
        t = trace or mt.trace(name="openai.chat")
        streaming = bool(kwargs.get("stream"))
        excluded = ("model", "messages", "stream", "stream_options") if streaming else ("model", "messages")
        gen = t.generation(
            name="openai.chat.completions",
            model=kwargs.get("model"),
            provider="openai",
            input=kwargs.get("messages"),
            modelParameters={k: v for k, v in kwargs.items() if k not in excluded},
        )

        if streaming:
            # Ask the API to include usage in the final chunk — never override an
            # explicit caller value.
            kwargs.setdefault("stream_options", {"include_usage": True})
            try:
                raw = original(*args, **kwargs)
            except Exception as e:  # noqa: BLE001
                gen.end(level="ERROR", statusMessage=str(e))
                raise

            acc = _ChatAccumulator()

            def on_chunk(chunk: Any) -> None:
                acc.add(chunk)

            def on_done(err: Any) -> None:
                output, usage = acc.finalize()
                if err is _ABANDONED:
                    gen.end(
                        level="WARNING", statusMessage="stream ended before completion", output=output, usage=usage
                    )
                elif err is not None:
                    gen.end(level="ERROR", statusMessage=str(err), output=output, usage=usage)
                else:
                    gen.end(output=output, usage=usage)

            return _RecordingStream(raw, on_chunk, on_done)

        try:
            resp = original(*args, **kwargs)
            gen.end(output=_message(resp), usage=_map_chat_usage(getattr(resp, "usage", None)))
            return resp
        except Exception as e:  # noqa: BLE001
            gen.end(level="ERROR", statusMessage=str(e))
            raise

    completions.create = create  # type: ignore[assignment]


class _ChatAccumulator:
    """Accumulates ``chat.completions`` streaming chunks (``choices[].delta``) into the
    same ``(output, usage)`` shape the non-streaming path builds."""

    def __init__(self) -> None:
        self._choices: dict[int, dict[str, Any]] = {}
        self._usage: Any = None

    def add(self, chunk: Any) -> None:
        usage = _get(chunk, "usage")
        if usage is not None:
            self._usage = usage
        for choice in _get(chunk, "choices", []) or []:
            index = _get(choice, "index", 0)
            slot = self._choices.setdefault(index, {"role": None, "content": None, "refusal": None, "tool_calls": {}})
            delta = _get(choice, "delta")
            if delta is None:
                continue
            role = _get(delta, "role")
            if role is not None:
                slot["role"] = role
            content = _get(delta, "content")
            if content:
                slot["content"] = (slot["content"] or "") + content
            refusal = _get(delta, "refusal")
            if refusal:
                slot["refusal"] = (slot["refusal"] or "") + refusal
            for tc in _get(delta, "tool_calls", []) or []:
                tc_index = _get(tc, "index", 0)
                tc_slot = slot["tool_calls"].setdefault(
                    tc_index, {"id": None, "type": None, "function": {"name": None, "arguments": ""}}
                )
                tc_id = _get(tc, "id")
                if tc_id is not None:
                    tc_slot["id"] = tc_id
                tc_type = _get(tc, "type")
                if tc_type is not None:
                    tc_slot["type"] = tc_type
                func = _get(tc, "function")
                if func is not None:
                    name = _get(func, "name")
                    if name is not None:
                        tc_slot["function"]["name"] = name
                    args = _get(func, "arguments")
                    if args:
                        tc_slot["function"]["arguments"] += args

    def finalize(self) -> tuple[Any, Optional[dict[str, Any]]]:
        if not self._choices:
            return None, _map_chat_usage(self._usage)
        index = 0 if 0 in self._choices else min(self._choices)
        slot = self._choices[index]
        message: dict[str, Any] = {"role": slot["role"], "content": slot["content"]}
        if slot["refusal"] is not None:
            message["refusal"] = slot["refusal"]
        if slot["tool_calls"]:
            message["tool_calls"] = [slot["tool_calls"][i] for i in sorted(slot["tool_calls"])]
        return message, _map_chat_usage(self._usage)


def _map_chat_usage(usage: Any) -> Optional[dict[str, Any]]:
    if usage is None:
        return None
    return {
        "promptTokens": _get(usage, "prompt_tokens"),
        "completionTokens": _get(usage, "completion_tokens"),
        "totalTokens": _get(usage, "total_tokens"),
    }


def _patch_responses(client: Any, mt: Memoturn, trace: Optional[Trace]) -> None:
    responses = getattr(client, "responses", None)
    if responses is None or not hasattr(responses, "create"):
        return  # older openai SDK without the Responses API
    original = responses.create

    def create(*args: Any, **kwargs: Any) -> Any:
        t = trace or mt.trace(name="openai.responses")
        instructions = kwargs.get("instructions")
        input_val = kwargs.get("input")
        streaming = bool(kwargs.get("stream"))
        excluded = (
            ("model", "input", "instructions", "stream", "stream_options")
            if streaming
            else ("model", "input", "instructions")
        )
        gen = t.generation(
            name="openai.responses",
            model=kwargs.get("model"),
            provider="openai",
            # Keep instructions (system-equivalent) alongside the input when present.
            input={"instructions": instructions, "input": input_val} if instructions is not None else input_val,
            modelParameters={k: v for k, v in kwargs.items() if k not in excluded},
        )

        if streaming:
            try:
                raw = original(*args, **kwargs)
            except Exception as e:  # noqa: BLE001
                gen.end(level="ERROR", statusMessage=str(e))
                raise

            output: Any = None
            usage: Optional[dict[str, Any]] = None
            terminal_seen = False
            failed = False

            def on_chunk(event: Any) -> None:
                nonlocal output, usage, terminal_seen, failed
                etype = _get(event, "type")
                if etype not in ("response.completed", "response.failed", "response.incomplete"):
                    return
                response = _get(event, "response")
                if response is not None:
                    output = _responses_output(response)
                    usage = _map_responses_usage(_get(response, "usage"))
                terminal_seen = True
                if etype == "response.failed":
                    failed = True

            def on_done(err: Any) -> None:
                if err is _ABANDONED:
                    gen.end(
                        level="WARNING", statusMessage="stream ended before completion", output=output, usage=usage
                    )
                elif err is not None:
                    gen.end(level="ERROR", statusMessage=str(err), output=output, usage=usage)
                elif failed:
                    gen.end(level="ERROR", statusMessage="response.failed", output=output, usage=usage)
                elif not terminal_seen:
                    gen.end(
                        level="ERROR",
                        statusMessage="stream ended before a terminal response event",
                        output=output,
                        usage=usage,
                    )
                else:
                    gen.end(output=output, usage=usage)

            return _RecordingStream(raw, on_chunk, on_done)

        try:
            resp = original(*args, **kwargs)
            gen.end(output=_responses_output(resp), usage=_map_responses_usage(getattr(resp, "usage", None)))
            return resp
        except Exception as e:  # noqa: BLE001
            gen.end(level="ERROR", statusMessage=str(e))
            raise

    responses.create = create  # type: ignore[assignment]


def _message(resp: Any) -> Any:
    try:
        return resp.choices[0].message.model_dump()
    except Exception:  # noqa: BLE001
        return str(resp)


def _responses_output(resp: Any) -> Any:
    text = _get(resp, "output_text")
    if text is not None:
        return text
    out = _get(resp, "output")
    if out is not None:
        try:
            return [o.model_dump() if hasattr(o, "model_dump") else o for o in out]
        except Exception:  # noqa: BLE001
            return str(out)
    return str(resp)


def _map_responses_usage(usage: Any) -> Optional[dict[str, Any]]:
    if usage is None:
        return None
    return {
        "promptTokens": _get(usage, "input_tokens"),
        "completionTokens": _get(usage, "output_tokens"),
        "totalTokens": _get(usage, "total_tokens"),
    }

File: sdks/js/src/openai.ts
import type { Memoturn, MemoturnTrace } from "./client.js";
import { tapStream } from "./stream.js";

/**
 * Drop-in wrapper for the OpenAI SDK. Wraps both `client.chat.completions.create` and
 * `client.responses.create` so each call is recorded as a memoturn generation (model,
 * params, usage, latency, output) — including streaming calls (`stream: true`), which are
 * accumulated into the same output/usage shape as a non-streaming call while still being
 * yielded to the caller in real time (no buffering, no added latency).
 *
 *   const openai = wrapOpenAI(new OpenAI(), memoturn);
 *   await openai.chat.completions.create({ model, messages });
 *   await openai.responses.create({ model, input });
 *
 * Pass `{ trace }` to nest generations under an existing trace; otherwise each call
 * gets its own trace. Pass `{ streamTimeoutMs }` to override the idle-stream abandonment
 * backstop (default 120s).
 */
export function wrapOpenAI<T extends object>(
  client: T,
  memoturn: Memoturn,
  options: { trace?: MemoturnTrace; traceName?: string; streamTimeoutMs?: number } = {},
): T {
  return new Proxy(client, {
    get(target, prop, receiver) {
      const value = Reflect.get(target, prop, receiver);
      if (prop === "chat") return wrapChat(value, memoturn, options);
      if (prop === "responses") return wrapResponses(value, memoturn, options);
      return value;
    },
  });
}

type WrapOptions = { trace?: MemoturnTrace; traceName?: string; streamTimeoutMs?: number };

function wrapChat(chat: any, memoturn: Memoturn, options: WrapOptions): any {
  return new Proxy(chat, {
    get(target, prop, receiver) {
      const value = Reflect.get(target, prop, receiver);
      if (prop === "completions") return wrapCompletions(value, memoturn, options);
      return value;
    },
  });
}

/** Merge streamed chat-completion chunks into the same `{ output, usage }` shape the
 * non-streaming path records (`response.choices[0].message` + mapped usage). */
function createChatStreamAccumulator() {
  const choices = new Map<
    number,
    {
      role?: string;
      content: string;
      refusal: string;
      toolCalls: Map<number, { id?: string; type?: string; function: { name?: string; arguments: string } }>;
    }
  >();
  let usage: any;

  return {
    add(chunk: any) {
      if (chunk?.usage) usage = chunk.usage;
      for (const choice of chunk?.choices ?? []) {
        const index = choice.index ?? 0;
        let entry = choices.get(index);
        if (!entry) {
          entry = { content: "", refusal: "", toolCalls: new Map() };
          choices.set(index, entry);
        }
        const delta = choice.delta ?? {};
        if (delta.role) entry.role = delta.role;
        if (typeof delta.content === "string") entry.content += delta.content;
        if (typeof delta.refusal === "string") entry.refusal += delta.refusal;
        for (const toolCall of delta.tool_calls ?? []) {
          const tcIndex = toolCall.index ?? 0;
          let tc = entry.toolCalls.get(tcIndex);
          if (!tc) {
            tc = { function: { arguments: "" } };
            entry.toolCalls.set(tcIndex, tc);
          }
          if (toolCall.id) tc.id = toolCall.id;
          if (toolCall.type) tc.type = toolCall.type;
          if (toolCall.function?.name) tc.function.name = toolCall.function.name;
          if (typeof toolCall.function?.arguments === "string") tc.function.arguments += toolCall.function.arguments;
        }
      }
    },
    finalize() {
      const indexes = [...choices.keys()].sort((a, b) => a - b);
      const messages = indexes.map((i) => {
        const entry = choices.get(i)!;
        const message: any = { role: entry.role ?? "assistant", content: entry.content };
        if (entry.refusal) message.refusal = entry.refusal;
        if (entry.toolCalls.size > 0) {
          message.tool_calls = [...entry.toolCalls.keys()]
            .sort((a, b) => a - b)
            .map((idx) => {
              const tc = entry.toolCalls.get(idx)!;
              return {
                id: tc.id,
                type: tc.type,
                function: { name: tc.function.name, arguments: tc.function.arguments },
              };
            });
        }
        return message;
      });
      return {
        // Match the non-streaming shape: a single message object when there's one choice
        // (the common case), mirroring `response?.choices?.[0]?.message`.
        output: messages.length <= 1 ? messages[0] : messages,
        usage: usage
          ? {
              promptTokens: usage.prompt_tokens,
              completionTokens: usage.completion_tokens,
              totalTokens: usage.total_tokens,
            }
          : undefined,
      };
    },
  };
}

function wrapCompletions(completions: any, memoturn: Memoturn, options: WrapOptions): any {
  return new Proxy(completions, {
    get(target, prop, receiver) {
      const original = Reflect.get(target, prop, receiver);
      if (prop !== "create" || typeof original !== "function") return original;

      return async function create(params: any, ...rest: any[]) {
        const trace = options.trace ?? memoturn.trace({ name: options.traceName ?? "openai.chat" });
        const { model, messages, stream, stream_options, ...modelParameters } = params ?? {};
        const generation = trace.generation({
          name: "openai.chat.completions",
          model,
          provider: "openai",
          modelParameters,
          input: messages,
        });

        if (stream) {
          // Ask for a final usage-bearing chunk unless the caller already made an explicit
          // choice (including opting out with `include_usage: false`) — never override that.
          const streamParams =
            stream_options === undefined ? { ...params, stream_options: { include_usage: true } } : params;
          try {
            const response = await original.call(target, streamParams, ...rest);
            const accumulator = createChatStreamAccumulator();
            return tapStream(
              response,
              {
                onChunk: (chunk: any) => accumulator.add(chunk),
                onDone: (err, reason) => {
                  if (reason === "error") {
                    generation.end({ level: "ERROR", statusMessage: String(err), ...accumulator.finalize() });
                  } else if (reason === "abandoned") {
                    generation.end({
                      level: "WARNING",
                      statusMessage: "stream ended before completion",
                      ...accumulator.finalize(),
                    });
                  } else {
                    generation.end({ ...accumulator.finalize() });
                  }
                },
              },
              { idleTimeoutMs: options.streamTimeoutMs },
            );
          } catch (err) {
            generation.end({ level: "ERROR", statusMessage: String(err) });
            throw err;
          }
        }

        try {
          const response = await original.call(target, params, ...rest);
          generation.end({
            output: response?.choices?.[0]?.message ?? response,
            usage: response?.usage
              ? {
                  promptTokens: response.usage.prompt_tokens,
                  completionTokens: response.usage.completion_tokens,
                  totalTokens: response.usage.total_tokens,
                }
              : undefined,
          });
          return response;
        } catch (err) {
          generation.end({ level: "ERROR", statusMessage: String(err) });
          throw err;
        }
      };
    },
  });
}

/** Extract the output from a (possibly streamed-to-completion) Responses API `Response`. */
function extractResponsesOutput(response: any): unknown {
  return response?.output_text ?? response?.output ?? response;
}

/** Map Responses API usage (`input_tokens`/`output_tokens`) to the recorded usage shape. */
function mapResponsesUsage(usage: any): Record<string, number> | undefined {
  return usage
    ? {
        promptTokens: usage.input_tokens,
        completionTokens: usage.output_tokens,
        totalTokens: usage.total_tokens,
      }
    : undefined;
}

/**
 * Wrap `responses.create` — OpenAI's Responses API. Request `input` (+ `instructions`)
 * becomes the generation input; the response's `output_text`/`output` items become the
 * output; `input_tokens`/`output_tokens` map to the usage fields. Without this, teams on
 * the Responses API get no generations recorded.
 *
 * Streaming (`stream: true`) is event-based: the terminal event (`response.completed` /
 * `response.failed` / `response.incomplete`) carries the full final `Response` object, which
 * is what gets recorded.
 */
function wrapResponses(responses: any, memoturn: Memoturn, options: WrapOptions): any {
  return new Proxy(responses, {
    get(target, prop, receiver) {
      const original = Reflect.get(target, prop, receiver);
      if (prop !== "create" || typeof original !== "function") return original;

      return async function create(params: any, ...rest: any[]) {
        const trace = options.trace ?? memoturn.trace({ name: options.traceName ?? "openai.responses" });
        const { model, input, instructions, stream, ...modelParameters } = params ?? {};
        const generation = trace.generation({
          name: "openai.responses",
          model,
          provider: "openai",
          modelParameters,
          // Keep `instructions` (system-equivalent) alongside the input when present.
          input: instructions != null ? { instructions, input } : input,
        });

        if (stream) {
          try {
            const response = await original.call(target, params, ...rest);
            let finalResponse: any;
            let sawTerminalEvent = false;
            let failed = false;
            return tapStream(
              response,
              {
                onChunk: (event: any) => {
                  if (event?.type === "response.completed" || event?.type === "response.incomplete") {
                    finalResponse = event.response;
                    sawTerminalEvent = true;
                  } else if (event?.type === "response.failed") {
                    finalResponse = event.response;
                    sawTerminalEvent = true;
                    failed = true;
                  }
                },
                onDone: (err, reason) => {
                  const output = extractResponsesOutput(finalResponse);
                  const usage = mapResponsesUsage(finalResponse?.usage);
                  if (reason === "abandoned") {
                    generation.end({
                      level: "WARNING",
                      statusMessage: "stream ended before completion",
                      output,
                      usage,
                    });
                  } else if (reason === "error" || failed || !sawTerminalEvent) {
                    generation.end({
                      level: "ERROR",
                      statusMessage: err
                        ? String(err)
                        : failed
                          ? "response.failed"
                          : "stream ended without a terminal event",
                      output,
                      usage,
                    });
                  } else {
                    generation.end({ output, usage });
                  }
                },
              },
              { idleTimeoutMs: options.streamTimeoutMs },
            );
          } catch (err) {
            generation.end({ level: "ERROR", statusMessage: String(err) });
            throw err;
          }
        }

        try {
          const response = await original.call(target, params, ...rest);
          generation.end({ output: extractResponsesOutput(response), usage: mapResponsesUsage(response?.usage) });
          return response;
        } catch (err) {
          generation.end({ level: "ERROR", statusMessage: String(err) });
          throw err;
        }
      };
    },
  });
}
